File: create_customer_excel.mjs
import fs from 'fs';
import ExcelJS from 'exceljs';

// Read the backup file
const backupPath = '/Users/mtaasisi/Downloads/database-backup-2025-07-30T12-11-11-519Z.json';
const backupData = JSON.parse(fs.readFileSync(backupPath, 'utf8'));

console.log('📊 Creating Excel file with customer data...');
console.log(`📈 Total customers: ${backupData.data.customers.length}`);

// Create a new workbook and worksheet
const workbook = new ExcelJS.Workbook();
const worksheet = workbook.addWorksheet('Customers');

// Define the columns
worksheet.columns = [
  { header: 'ID', key: 'id', width: 40 },
  { header: 'Name', key: 'name', width: 20 },
  { header: 'Email', key: 'email', width: 25 },
  { header: 'Phone', key: 'phone', width: 15 },
  { header: 'Gender', key: 'gender', width: 10 },
  { header: 'City', key: 'city', width: 15 },
  { header: 'Joined Date', key: 'joined_date', width: 15 },
  { header: 'Loyalty Level', key: 'loyalty_level', width: 12 },
  { header: 'Color Tag', key: 'color_tag', width: 12 },
  { header: 'Total Spent', key: 'total_spent', width: 12 },
  { header: 'Points', key: 'points', width: 10 },
  { header: 'Last Visit', key: 'last_visit', width: 15 },
  { header: 'Is Active', key: 'is_active', width: 10 },
  { header: 'WhatsApp', key: 'whatsapp', width: 15 },
  { header: 'Referral Source', key: 'referral_source', width: 15 },
  { header: 'Birth Month', key: 'birth_month', width: 12 },
  { header: 'Birth Day', key: 'birth_day', width: 10 },
  { header: 'Customer Tag', key: 'customer_tag', width: 12 },
  { header: 'Notes', key: 'notes', width: 30 },
  { header: 'Total Returns', key: 'total_returns', width: 12 },
  { header: 'Initial Notes', key: 'initial_notes', width: 30 },
  { header: 'Location Description', key: 'location_description', width: 25 },
  { header: 'National ID', key: 'national_id', width: 15 },
  { header: 'Created At', key: 'created_at', width: 15 },
  { header: 'Updated At', key: 'updated_at', width: 15 }
];

// Style the header row
worksheet.getRow(1).font = { bold: true };
worksheet.getRow(1).fill = {
  type: 'pattern',
  pattern: 'solid',
  fgColor: { argb: 'FFE0E0E0' }
};

// Add data rows
backupData.data.customers.forEach((customer, index) => {
  const row = worksheet.addRow({
    id: customer.id,
    name: customer.name || '',
    email: customer.email || '',
    phone: customer.phone || '',
    gender: customer.gender || '',
    city: customer.city || '',
    joined_date: customer.joined_date ? new Date(customer.joined_date).toLocaleDateString() : '',
    loyalty_level: customer.loyalty_level || '',
    color_tag: customer.color_tag || '',
    total_spent: customer.total_spent || 0,
    points: customer.points || 0,
    last_visit: customer.last_visit ? new Date(customer.last_visit).toLocaleDateString() : '',
    is_active: customer.is_active ? 'Yes' : 'No',
    whatsapp: customer.whatsapp || '',
    referral_source: customer.referral_source || '',
    birth_month: customer.birth_month || '',
    birth_day: customer.birth_day || '',
    customer_tag: customer.customer_tag || '',
    notes: customer.notes || '',
    total_returns: customer.total_returns || 0,
    initial_notes: customer.initial_notes || '',
    location_description: customer.location_description || '',
    national_id: customer.national_id || '',
    created_at: customer.created_at ? new Date(customer.created_at).toLocaleDateString() : '',
    updated_at: customer.updated_at ? new Date(customer.updated_at).toLocaleDateString() : ''
  });

  // Add conditional formatting for important fields
  if (customer.whatsapp) {
    row.getCell('whatsapp').fill = {
      type: 'pattern',
      pattern: 'solid',
      fgColor: { argb: 'FFE8F5E8' } // Light green
    };
  }

  if (customer.referral_source) {
    row.getCell('referral_source').fill = {
      type: 'pattern',
      pattern: 'solid',
      fgColor: { argb: 'FFF0E8FF' } // Light purple
    };
  }

  if (customer.birth_month || customer.birth_day) {
    row.getCell('birth_month').fill = {
      type: 'pattern',
      pattern: 'solid',
      fgColor: { argb: 'FFFFE8F0' } // Light pink
    };
    row.getCell('birth_day').fill = {
      type: 'pattern',
      pattern: 'solid',
      fgColor: { argb: 'FFFFE8F0' } // Light pink
    };
  }

  // Highlight customers with high points
  if (customer.points > 100) {
    row.getCell('points').fill = {
      type: 'pattern',
      pattern: 'solid',
      fgColor: { argb: 'FFFFF2CC' } // Light yellow
    };
  }
});

// Add a summary worksheet
const summarySheet = workbook.addWorksheet('Summary');
summarySheet.columns = [
  { header: 'Metric', key: 'metric', width: 25 },
  { header: 'Count', key: 'count', width: 15 },
  { header: 'Percentage', key: 'percentage', width: 15 }
];

// Calculate statistics
const totalCustomers = backupData.data.customers.length;
const customersWithBirthdays = backupData.data.customers.filter(c => c.birth_month || c.birth_day).length;
const customersWithReferralSource = backupData.data.customers.filter(c => c.referral_source && c.referral_source !== '').length;
const customersWithWhatsApp = backupData.data.customers.filter(c => c.whatsapp && c.whatsapp !== '').length;
const customersWithHighPoints = backupData.data.customers.filter(c => c.points > 100).length;

// Add summary data
summarySheet.addRow({ metric: 'Total Customers', count: totalCustomers, percentage: '100%' });
summarySheet.addRow({ metric: 'Customers with Birthdays', count: customersWithBirthdays, percentage: `${((customersWithBirthdays / totalCustomers) * 100).toFixed(1)}%` });
summarySheet.addRow({ metric: 'Customers with Referral Source', count: customersWithReferralSource, percentage: `${((customersWithReferralSource / totalCustomers) * 100).toFixed(1)}%` });
summarySheet.addRow({ metric: 'Customers with WhatsApp', count: customersWithWhatsApp, percentage: `${((customersWithWhatsApp / totalCustomers) * 100).toFixed(1)}%` });
summarySheet.addRow({ metric: 'Customers with 100+ Points', count: customersWithHighPoints, percentage: `${((customersWithHighPoints / totalCustomers) * 100).toFixed(1)}%` });

// Style summary sheet
summarySheet.getRow(1).font = { bold: true };
summarySheet.getRow(1).fill = {
  type: 'pattern',
  pattern: 'solid',
  fgColor: { argb: 'FFE0E0E0' }
};

// Add referral source breakdown
const referralSheet = workbook.addWorksheet('Referral Sources');
referralSheet.columns = [
  { header: 'Referral Source', key: 'source', width: 20 },
  { header: 'Count', key: 'count', width: 15 },
  { header: 'Percentage', key: 'percentage', width: 15 }
];

// Calculate referral source breakdown
const referralSources = {};
backupData.data.customers.forEach(customer => {
  if (customer.referral_source && customer.referral_source !== '') {
    referralSources[customer.referral_source] = (referralSources[customer.referral_source] || 0) + 1;
  }
});

Object.entries(referralSources)
  .sort(([,a], [,b]) => b - a)
  .forEach(([source, count]) => {
    referralSheet.addRow({
      source,
      count,
      percentage: `${((count / totalCustomers) * 100).toFixed(1)}%`
    });
  });

// Style referral sheet
referralSheet.getRow(1).font = { bold: true };
referralSheet.getRow(1).fill = {
  type: 'pattern',
  pattern: 'solid',
  fgColor: { argb: 'FFE0E0E0' }
};

// Add birthday breakdown
const birthdaySheet = workbook.addWorksheet('Birthdays');
birthdaySheet.columns = [
  { header: 'Month', key: 'month', width: 15 },
  { header: 'Count', key: 'count', width: 15 },
  { header: 'Percentage', key: 'percentage', width: 15 }
];

// Calculate birthday breakdown
const birthMonths = {};
backupData.data.customers.forEach(customer => {
  if (customer.birth_month) {
    birthMonths[customer.birth_month] = (birthMonths[customer.birth_month] || 0) + 1;
  }
});

Object.entries(birthMonths)
  .sort(([,a], [,b]) => b - a)
  .forEach(([month, count]) => {
    birthdaySheet.addRow({
      month,
      count,
      percentage: `${((count / customersWithBirthdays) * 100).toFixed(1)}%`
    });
  });

// Style birthday sheet
birthdaySheet.getRow(1).font = { bold: true };
birthdaySheet.getRow(1).fill = {
  type: 'pattern',
  pattern: 'solid',
  fgColor: { argb: 'FFE0E0E0' }
};

// Save the workbook
const excelPath = './customer_data_export.xlsx';
await workbook.xlsx.writeFile(excelPath);

console.log('✅ Excel file created successfully!');
console.log(`📁 File saved as: ${excelPath}`);
console.log('\n📊 Summary:');
console.log(`📈 Total customers: ${totalCustomers}`);
console.log(`🎂 Customers with birthdays: ${customersWithBirthdays}`);
console.log(`📱 Customers with referral source: ${customersWithReferralSource}`);
console.log(`💬 Customers with WhatsApp: ${customersWithWhatsApp}`);
console.log(`💰 Customers with 100+ points: ${customersWithHighPoints}`);

console.log('\n📋 Excel file contains:');
console.log('1. Customers - Main data with all customer information');
console.log('2. Summary - Key statistics and percentages');
console.log('3. Referral Sources - Breakdown by referral source');
console.log('4. Birthdays - Breakdown by birth month');

console.log('\n🎯 How to use this Excel file:');
console.log('1. Open the file in Excel or Google Sheets');
console.log('2. Review the data in the "Customers" sheet');
console.log('3. Use the "Summary" sheet for quick statistics');
console.log('4. Use the "Referral Sources" and "Birthdays" sheets for marketing insights');
console.log('5. Update any missing or incorrect data directly in Excel');
console.log('6. Import the updated data back to your system'); 

File: create_customer_excel.py
#!/usr/bin/env python3
import pandas as pd
import re
import sys
from datetime import datetime

def parse_customer_data(filename):
    """Parse customer data from the backup file"""
    customers = []
    
    with open(filename, 'r', encoding='utf-8') as file:
        content = file.read()
    
    # Find the COPY statement and extract data
    copy_match = re.search(r'COPY public\.customers.*?FROM stdin;\n(.*?)\n\\.', content, re.DOTALL)
    if not copy_match:
        print("Could not find customer data in the file")
        return []
    
    data_section = copy_match.group(1)
    
    # Split by lines and parse each customer
    lines = data_section.strip().split('\n')
    
    for line in lines:
        if not line.strip():
            continue
            
        # Split by tab character
        parts = line.split('\t')
        if len(parts) < 28:  # Expected number of columns
            continue
            
        try:
            customer = {
                'ID': parts[0].strip(),
                'Name': parts[1].strip(),
                'Email': parts[2].strip() if parts[2].strip() != '\\N' else '',
                'Phone': parts[3].strip(),
                'Gender': parts[4].strip() if parts[4].strip() != '\\N' else '',
                'City': parts[5].strip() if parts[5].strip() != '\\N' else '',
                'Location Description': parts[6].strip() if parts[6].strip() != '\\N' else '',
                'National ID': parts[7].strip() if parts[7].strip() != '\\N' else '',
                'Joined Date': parts[8].strip() if parts[8].strip() != '\\N' else '',
                'Loyalty Level': parts[9].strip(),
                'Color Tag': parts[10].strip(),
                'Referred By': parts[11].strip() if parts[11].strip() != '\\N' else '',
                'Total Spent': parts[12].strip(),
                'Points': parts[13].strip(),
                'Last Visit': parts[14].strip() if parts[14].strip() != '\\N' else '',
                'Is Active': parts[15].strip(),
                'Referrals': parts[16].strip(),
                'WhatsApp': parts[17].strip() if parts[17].strip() != '\\N' else '',
                'Referral Source': parts[18].strip() if parts[18].strip() != '\\N' else '',
                'Birth Month': parts[19].strip() if parts[19].strip() != '\\N' else '',
                'Birth Day': parts[20].strip() if parts[20].strip() != '\\N' else '',
                'Customer Tag': parts[21].strip() if parts[21].strip() != '\\N' else '',
                'Notes': parts[22].strip() if parts[22].strip() != '\\N' else '',
                'Total Returns': parts[23].strip(),
                'Profile Image': parts[24].strip() if parts[24].strip() != '\\N' else '',
                'Created At': parts[25].strip() if parts[25].strip() != '\\N' else '',
                'Updated At': parts[26].strip() if parts[26].strip() != '\\N' else '',
                'Initial Notes': parts[27].strip() if parts[27].strip() != '\\N' else '',
                'Created By': parts[28].strip() if len(parts) > 28 and parts[28].strip() != '\\N' else ''
            }
            customers.append(customer)
        except Exception as e:
            print(f"Error parsing line: {line[:100]}... Error: {e}")
            continue
    
    return customers

def create_excel_file(customers, output_filename):
    """Create Excel file with customer data"""
    if not customers:
        print("No customer data to export")
        return
    
    # Create DataFrame
    df = pd.DataFrame(customers)
    
    # Reorder columns for better readability
    column_order = [
        'Name', 'Phone', 'Email', 'WhatsApp', 'Gender', 'City', 
        'Loyalty Level', 'Points', 'Total Spent', 'Is Active',
        'Referral Source', 'Referred By', 'Joined Date', 'Last Visit',
        'Customer Tag', 'Notes', 'Birth Month', 'Birth Day',
        'ID', 'Created At', 'Updated At'
    ]
    
    # Only include columns that exist in the data
    available_columns = [col for col in column_order if col in df.columns]
    df = df[available_columns]
    
    # Create Excel writer with formatting
    with pd.ExcelWriter(output_filename, engine='openpyxl') as writer:
        # Write main data
        df.to_excel(writer, sheet_name='All Customers', index=False)
        
        # Get the workbook and worksheet
        workbook = writer.book
        worksheet = writer.sheets['All Customers']
        
        # Auto-adjust column widths
        for column in worksheet.columns:
            max_length = 0
            column_letter = column[0].column_letter
            
            for cell in column:
                try:
                    if len(str(cell.value)) > max_length:
                        max_length = len(str(cell.value))
                except:
                    pass
            
            adjusted_width = min(max_length + 2, 50)
            worksheet.column_dimensions[column_letter].width = adjusted_width
        
        # Add summary sheet
        summary_data = {
            'Metric': [
                'Total Customers',
                'Active Customers',
                'Inactive Customers',
                'Customers with Email',
                'Customers with WhatsApp',
                'Customers with Points > 0',
                'Bronze Level Customers',
                'Silver Level Customers',
                'Gold Level Customers',
                'Platinum Level Customers',
                'Top Cities',
                'Top Referral Sources'
            ],
            'Count': [
                len(df),
                len(df[df['Is Active'] == 't']),
                len(df[df['Is Active'] == 'f']),
                len(df[df['Email'] != '']),
                len(df[df['WhatsApp'] != '']),
                len(df[df['Points'].astype(int) > 0]),
                len(df[df['Loyalty Level'] == 'bronze']),
                len(df[df['Loyalty Level'] == 'silver']),
                len(df[df['Loyalty Level'] == 'gold']),
                len(df[df['Loyalty Level'] == 'platinum']),
                df['City'].value_counts().head(3).to_string(),
                df['Referral Source'].value_counts().head(5).to_string()
            ]
        }
        
        summary_df = pd.DataFrame(summary_data)
        summary_df.to_excel(writer, sheet_name='Summary', index=False)
        
        # Format summary sheet
        summary_worksheet = writer.sheets['Summary']
        for column in summary_worksheet.columns:
            max_length = 0
            column_letter = column[0].column_letter
            
            for cell in column:
                try:
                    if len(str(cell.value)) > max_length:
                        max_length = len(str(cell.value))
                except:
                    pass
            
            adjusted_width = min(max_length + 2, 50)
            summary_worksheet.column_dimensions[column_letter].width = adjusted_width
    
    print(f"✅ Excel file created successfully: {output_filename}")
    print(f"📊 Total customers exported: {len(customers)}")

def main():
    backup_file = "/Users/mtaasisi/Desktop/SQL/supabase_full_2025-08-08_09-48.sql"
    output_file = "LATS_Customer_Contacts.xlsx"
    
    print("🔄 Parsing customer data from backup...")
    customers = parse_customer_data(backup_file)
    
    if customers:
        print(f"📋 Found {len(customers)} customers")
        print("📊 Creating Excel file...")
        create_excel_file(customers, output_file)
        
        # Show some sample data
        print("\n📋 Sample customers:")
        for i, customer in enumerate(customers[:5]):
            print(f"{i+1}. {customer['Name']} - {customer['Phone']} - {customer['City']}")
    else:
        print("❌ No customer data found")

if __name__ == "__main__":
    main()
